File: CarHomeWidget/src/com/anod/car/home/ui/WidgetsListLoader.java
package com.anod.car.home.ui;

import android.content.Context;
import android.support.v4.content.AsyncTaskLoader;
import android.util.SparseArray;
import com.anod.car.home.model.LauncherShortcutsModel;
import com.anod.car.home.model.ShortcutInfo;

import java.util.HashMap;

/**
 * @author alex
 * @date 5/27/13
 */
public class WidgetsListLoader extends AsyncTaskLoader< HashMap<Integer,SparseArray<ShortcutInfo>> > {
	private final Context mContext;
	private final int[] mAppWidgetIds;

	public WidgetsListLoader(Context context, int[] appWidgetIds) {
		super(context);
		mContext = context;
		mAppWidgetIds = appWidgetIds;
	}

	@Override
	public HashMap<Integer,SparseArray<ShortcutInfo>> loadInBackground() {

		HashMap<Integer,SparseArray<ShortcutInfo>> result = new HashMap<Integer, SparseArray<ShortcutInfo>>();

		for (int i = 0; i < mAppWidgetIds.length; i++) {
			LauncherShortcutsModel model = new LauncherShortcutsModel(mContext,mAppWidgetIds[i]);
			model.init();

			result.put(mAppWidgetIds[i], model.getShortcuts());
		}

		return result;
	}

}

File: CarHomeWidget/src/main/java/com/anod/car/home/ui/WidgetsListLoader.kt
package com.anod.car.home.ui

import com.anod.car.home.appwidget.WidgetHelper
import com.anod.car.home.model.Shortcut
import com.anod.car.home.model.WidgetShortcutsModel

import android.content.Context
import android.support.v4.content.AsyncTaskLoader
import android.util.SparseArray

/**
 * @author alex
 * @date 5/27/13
 */
class WidgetsListLoader(context: Context) : AsyncTaskLoader<WidgetsListLoader.Result>(context) {

    class Result {
        var large: SparseArray<SparseArray<Shortcut>> = SparseArray()
        var shortcuts: IntArray = intArrayOf()
    }

    override fun loadInBackground(): Result {

        val appWidgetIds = WidgetHelper.getLargeWidgetIds(context)
        val result = Result()

        for (i in appWidgetIds.indices) {
            val model = WidgetShortcutsModel(context, appWidgetIds[i])
            model.init()

            result.large.put(appWidgetIds[i], model.shortcuts)
        }

        result.shortcuts = WidgetHelper.getShortcutWidgetIds(context)
        return result
    }

}
